README.md
[![Gem Version](https://badge.fury.io/rb/urlenc.svg)](https://badge.fury.io/rb/urlenc)

# Urlenc

The simplest way to URL encode / decode from STDIN. URL encoding is a generic name, more accurately called Percent-encoding. The URL encoding is by far the most well-known name, so we intentionally named this gem Urlenc.

See about Percent-encoding: [wikipedia](https://en.wikipedia.org/wiki/Percent-encoding)

## Installation

Just install from rubygems:

    $ gem install urlenc

## Usage

```bash
$ # decode
$ echo "http://sample%20Text%21/" | urlenc -d
http://sample Text!/

$ # encode with split by /
$ echo "http://sample Text!/" | urlenc -es
http://sample%20Text%21/

$ # or just encode
$ echo "http://sample Text!/" | urlenc -e
http%3A%2F%2Fsample%20Text%21%2F
```

```
Usage: urlenc {decode/encode} [options]
 Required:
    -d, --decode    Decode/unescape STDIN. %22Sample%20Text%21%22 to "Sample Text!"
    -e, --encode    Encode/escape   STDIN. "Sample Text!" to %22Sample%20Text%21%22

 Optional:
    -s, --split     Split before Encode/escape. http://sample Text!/ to http://sample%20Text%21/

 Common options:
    -h, --help      Show this message
    -V, --version   Show version
```

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome!
1. Fork it ( https://github.com/shinyaohtani/urlenc )
1. Create your feature branch (git checkout -b my-new-feature)
1. Commit your changes (git commit -am 'Add some feature')
1. Push to the branch (git push origin my-new-feature)
1. Create a new Pull Request


## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the Urlenc project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/[USERNAME]/urlenc/blob/master/CODE_OF_CONDUCT.md).


CHANGELOG.md
# Changelog

## [v1.0.1](https://github.com/shinyaohtani/urlenc/tree/v1.0.1) (2020-05-21)

[Full Changelog](https://github.com/shinyaohtani/urlenc/compare/v1.0.0...v1.0.1)

- Encoding is default behavior without specify "-e or -d"

**Implemented enhancements:**

- Is it better to treat no params as "-e"? [\#3](https://github.com/shinyaohtani/urlenc/issues/3)

**Merged pull requests:**

- If no params are given, try encoding. [\#4](https://github.com/shinyaohtani/urlenc/pull/4) ([shinyaohtani](https://github.com/shinyaohtani))

# Changelog

## [v1.0.0](https://github.com/shinyaohtani/urlenc/tree/v1.0.0) (2020-05-20)

[Full Changelog](https://github.com/shinyaohtani/urlenc/compare/58edee48598a45d6b4cbb74c12058f99aa96f97c...v1.0.0)

- first release



\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


spec/spec_helper.rb
# frozen_string_literal: true

require 'bundler/setup'
require 'urlenc'

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = '.rspec_status'

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end


spec/urlenc_spec.rb
# frozen_string_literal: true

RSpec.describe Urlenc do
  it 'has a version number' do
    expect(Urlenc::VERSION).not_to be nil
  end

  #:it "does something useful" do
  #:  expect(false).to eq(true)
  #:end
end


lib/urlenc.rb
# frozen_string_literal: true

require 'urlenc/version'

# Usage: urlenc [options]
#     -d, --decode  : Decode/unescape STDIN
#     -e, --encode  : Encode/escape   STDIN
module Urlenc
  require 'erb'
  require 'optparse'

  DESC_DECODE = <<~DECODE
    Decode/unescape STDIN. %22Sample%20Text%21%22 will be
    converted to "Sample Text!"
  DECODE
  DESC_ENCODE = <<~ENCODE
    Encode/escape   STDIN. "Sample Text!" will be converted
    to %22Sample%20Text%21%22
    *default: --encode
  ENCODE
  DESC_SPLIT = <<~SPLIT
    Split before Encode/escape. http://sample Text!/ will
    be converted to http://sample%20Text%21/
    *default: --no-split
  SPLIT
  # Usage: urlenc [options]
  #     -d, --decode  : Decode/unescape STDIN
  #     -e, --encode  : Encode/escape   STDIN
  class Urlenc
    attr_accessor :params

    def initialize
      @params = { encode: true }
    end

    def parse_options
      OptionParser.new do |opts|
        opts = define_options(opts)
        opts.parse!(ARGV, into: @params)
        puts opts.help if @params[:decode].nil? && @params[:encode].nil?
      end
    end

    def run
      if @params[:decode]
        STDIN.each { |line| puts CGI.unescape(line.chomp) }
      elsif @params[:split]
        STDIN.each do |line|
          protocol = line.chomp!.slice!(%r{^.+://}) || ''
          puts protocol + line.split('/', -1).map { |e| ERB::Util.url_encode(e) }.join('/')
        end
      else
        STDIN.each { |line| puts ERB::Util.url_encode(line.chomp) }
      end
    end

    def define_options(opts) # rubocop:disable Metrics/AbcSize
      opts.version = VERSION
      opts.separator ' Required:'
      opts.on('-d', '--decode', *DESC_DECODE.split(/\R/)) { |v| v }
      opts.on('-e', '--encode', *DESC_ENCODE.split(/\R/)) { |v| v }
      opts.separator ''
      opts.separator ' Optional:'
      opts.on('-s', '--[no-]split', *DESC_SPLIT.split(/\R/)) { |v| v }
      opts.separator ''
      opts.separator ' Common options:'
      opts.on_tail('-h', '--help', 'Show this message') do
        puts opts
        exit
      end
      opts.on_tail('-V', '--version', 'Show version') do
        puts opts.ver
        exit
      end
      opts.banner = <<~BANNER

        #{opts.ver}
        #{DESCRIPTION}
        Usage: #{opts.program_name} {decode/encode} [options]
      BANNER
      opts
    end
  end
end
